chp6 boggle problem

sj43 opened the pull request:
This is a pure recursive solution that yields TLE on Algospot.
Dynamic programming solution necessary to AC.

sj43 committed:
chp6 boggle problem

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+dx = [-1, -1, -1, 1, 1, 1, 0, 0]
+dy = [-1, 0, 1, -1, 0, 1, -1, 1]
+
+def inRange(y, x):
```
isRange -> is_range

Java = camelCase. e.g. isFound?
python = under_score. e.g. is_found?

this convention is used for both function and variable names

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+    return True
+
+
+def hasWord(y, x, word):
```
same here. has_word

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def hasWord(y, x, word):
+    if not inRange(y, x):
```
same.

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+    N = input()
+    for l in range(N):
+        s = raw_input()
+        alreadyfound = False
```
already_found

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+for _ in xrange(C):
+    for i in xrange(5):
+        a = list(raw_input())
+        board.append(a)
```
board.append(raw_input().strip()) ? because a is never used. raw_input() probably has a trailing '\n'. e.g. "URLPM\n"

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+
+C = input()
+for _ in xrange(C):
+    for i in xrange(5):
```
i is not used anywhere

for _ in xrange(5)

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+        a = list(raw_input())
+        board.append(a)
+    N = input()
+    for l in range(N):
```
for _ in xrange(N):

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+    for l in range(N):
+        s = raw_input()
+        alreadyfound = False
+        for j in xrange(5):
```
use meaningful name y.

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+        for j in xrange(5):
+            if alreadyfound:
+                break
+            for k in xrange(5):
```
use meaningful name x

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+            for k in xrange(5):
+                if alreadyfound:
+                    break
+                if hasWord(j, k, s):
```
if has_word(y, x, s)

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,47 @@
+        board.append(a)
+    N = input()
+    for l in range(N):
+        s = raw_input()
```
use meaningful name.

word_to_find?

sj43 committed:
Update boggle.py

comments resolved

sj43 committed:
BOGGLE - DP solution

Need feedback on how to follow C++
convention of using global cache (to keep function parameter as similar as possible to C++ function parameters)

sj43 commented:
DP solution: AC (python 2.7) on Algospot

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,59 @@
+
+def has_word(y, x, word, cur_index, cache):
+    if not in_range(y, x):
+        return False
```
add an empty line after line14 for visual clarity?

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,59 @@
+    cache[y][x][cur_index] = 1
+    if board[y][x] != word[cur_index]:
+        return False
+    if cur_index == len(word)-1:
```
always add spaces around a binary operator
e.g.
1 + 2
len(word) - 1
y + dy[direction]

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,59 @@
+    if board[y][x] != word[cur_index]:
+        return False
+    if cur_index == len(word)-1:
+        return True
```
add a line?

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,59 @@
+    if cur_index == len(word)-1:
+        return True
+    for direction in xrange(8):
+        nextY = y + dy[direction]
```
next_y
next_x

riceluxs1t replied:
we want the underscore convention instead of the camel case one

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,59 @@
+        nextX = x + dx[direction]
+        if not in_range(nextY, nextX):
+            continue
+        if cache[nextY][nextX][cur_index+1]:
```
cur_index + 1

riceluxs1t replied:
make sure to change here as well when you address my comment on cache

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,59 @@
+            continue
+        if cache[nextY][nextX][cur_index+1]:
+            continue
+        if has_word(nextY, nextX, word, cur_index+1, cache):
```
cur_index + 1

riceluxs1t commented on ch6/boggle.py:
```diff
@@ -0,0 +1,59 @@
+    return False
+
+
+def initialize_cache():
```
I think a more pythonic solution is to use a 3-tuple.

e.g.

cache[(y, x, cur_index)] = ...

instead of

cache[y][x][cur_index]

In a language like C, allocating a fixed size data structure is more natural.
in a language like Python, allocating a dynamically growing data structure is more natural.

The omitted part of the C(++) solution most likely initializes a cache of some fixed size
e.g.
int cache[UPPER_BOUND_ON_Y][UPPER_BOUND_ON_X][UPPER_BOUND_ON_INDEX];

where UPPER_BOUND_ON_XXX is the maximum possible of the corresponding dimension.

riceluxs1t reviewed (commented):
looks very clean! address those comments and then go ahead and merge it in yourself :) gj

sj43 committed:
ch6 boggle

Comments resolved. Merging.
